lib/ansible/module_utils/network/fortimanager/common.py

```diff
@@ -0,0 +1,317 @@
+        },
+        -10131: {
+            "msg": "Object dependency failed. Do all named objects in parameters exist?",
+            "changed": False,
+            "skipped": True
+        },
+        -9998: {
+            "msg": "Duplicate object. Try using mode='set', if using add. STOPPING. Use 'ignore_errors=yes' in playbook"
+            "to override and mark successful.",
+        },
+        -20042: {
+            "msg": "Device Unreachable.",
+            "skipped": True
+        },
+        -10033: {
+            "msg": "Duplicate object. Try using mode='set', if using add.",
+            "changed": False,
+            "skipped": True
+        },
+        -10000: {
+            "msg": "Duplicate object. Try using mode='set', if using add.",
+            "changed": False,
+            "skipped": True
+        },
+        -20010: {
+            "msg": "Device already added to FortiManager. Serial number already in use.",
+            "good_codes": [0, -20010],
+            "changed": False,
+            "stop_on_success": True
+        },
+        -20002: {
+            "msg": "Invalid Argument -- Does this Device exist on FortiManager?",
+            "changed": False,
+            "skipped": True,
+        }
+    }
+}
+
+DEFAULT_RESULT_OBJ = (-100000, {"msg": "Nothing Happened. Check that handle_response is being called!"})
+FAIL_SOCKET_MSG = {"msg": "Socket Path Empty! The persistent connection manager is messed up. "
+                   "Try again in a few moments."}
+
+
+# BEGIN ERROR EXCEPTIONS
+class FMGBaseException(Exception):
+    """Wrapper to catch the unexpected"""
+
+    def __init__(self, msg=None, *args, **kwargs):
+        if msg is None:
+            msg = "An exception occurred within the fortimanager.py httpapi connection plugin."
+        super(FMGBaseException, self).__init__(msg, *args)
+
+
+class FMGValidSessionException(FMGBaseException):
+    """Raised when a call is made, but there is no valid login instance"""
+
+    def __init__(self, method, params, *args, **kwargs):
+        msg = "A call using the {method} method was requested to {url} on a FortiManager instance that had no " \
+              "valid session or was not connected. Paramaters were:\n{params}". \
+            format(method=method, url=params[0]["url"], params=params)
+        super(FMGValidSessionException, self).__init__(msg, *args, **kwargs)
+
+
+class FMGValueError(ValueError):
+    """Catch value errors such as bad timeout values"""
+
+    def __init__(self, *args):
+        super(FMGValueError, self).__init__(*args)
+
+
+class FMGResponseNotFormedCorrect(KeyError):
+    """Used only if a response does not have a standard format as based on FMG response guidelines"""
+
+    def __init__(self, *args):
+        super(FMGResponseNotFormedCorrect, self).__init__(*args)
+
+# END ERROR CLASSES
+
+
+# BEGIN CLASSES
+class FMGRCommon(object):
+
+    @staticmethod
+    def format_request(method, url, *args, **kwargs):
```

Comment:
Why does this function not live in `plugins/httpapi/fortimanager`? Is it expected to be useful outside of the plugin?

Reply:
It is used by module_utils before being sent to the plugin. We do expect to use this outside of the plugin in the future as well. It was also an effort to keep the plugin as small as possible.